File: Minimize_the_Maximum_Difference_of_Pairs.cpp
class Solution {
public:
    int minimizeMax(vector<int>& nums, int p) {
        sort(nums.begin(), nums.end());

        auto isFeasible = [&](int diff) {
            int count = 0;
            int i = 0;
            while (i < nums.size() - 1 && count < p) {
                if (nums[i + 1] - nums[i] <= diff) {
                    count++;
                    i += 2;
                } else {
                    i++;
                }
            }
            return count >= p;
        };

        int left = 0;
        int right = nums[nums.size() - 1] - nums[0];

        while (left < right) {
            int midDiff = (left + right) / 2;
            if (isFeasible(midDiff)) {
                right = midDiff;
            } else {
                left = midDiff + 1;
            }
        }

        return left;
    }
};

File: Minimize_the_Maximum_Difference_of_Pairs.js
/**
 * @param {number[]} nums
 * @param {number} p
 * @return {number}
 */
var minimizeMax = function(nums, p) {
    nums.sort((a, b) => a - b);

    function isFeasible(diff) {
        let count = 0;
        let i = 0;
        while (i < nums.length - 1 && count < p) {
            if (nums[i + 1] - nums[i] <= diff) {
                count++;
                i += 2;
            } else {
                i++;
            }
        }
        return count >= p;
    }

    let left = 0;
    let right = nums[nums.length - 1] - nums[0];

    while (left < right) {
        let midDiff = Math.floor((left + right) / 2);
        if (isFeasible(midDiff)) {
            right = midDiff;
        } else {
            left = midDiff + 1;
        }
    }

    return left;    
};
